### ISSUE TITLE
Spread operator on Optional

### ISSUE THREAD

**Issue by Gregoirevda** [reasonml-community/error-message-improvement#27]
```reason
type book = {
  title: string
  description: string
};

type state = {
  book: option book
};

let book = {
  title: "e",
  description: "e"
};

let state = {
 book: Some book
};

let newState newDescription = {...state, book: {...state.book, description: newDescription }}
```

Should output: trying to deconstruct an Optional => Forbidden!

**Comment by bassjacob** [comment]
and some context from the discord: 
![image](https://user-images.githubusercontent.com/3908721/31099968-7add0f94-a813-11e7-83d8-a10ad2b86fe6.png)

**Comment by Octachron** [comment]
More precisely, this bug was fixed in OCaml 4.05 (with https://github.com/ocaml/ocaml/pull/901) . Nevertheless, any code manifesting this bug also raises a warning 23(all fields of a record are overwritten, thus the with clause is pointless), so turning this warning (with the equivalent of `-w @23`) to an error may be a workaround at the cost of forbidding variations on
```
type t = { x: int };
let x = { ...{x:0}, x : 1 };
```
